multiplex.go

```diff
@@ -104,41 +110,45 @@ func NewMultiplex(con net.Conn, initiator bool, memoryManager MemoryManager) (*M
 	mp := &Multiplex{
 		con:           con,
 		initiator:     initiator,
-		buf:           bufio.NewReader(con),
 		channels:      make(map[streamID]*Stream),
 		closed:        make(chan struct{}),
 		shutdown:      make(chan struct{}),
-		writeCh:       make(chan []byte, 16),
 		nstreams:      make(chan *Stream, 16),
 		memoryManager: memoryManager,
 	}
 
-	// up-front reserve memory for max buffers
-	bufs := 0
-	var err error
-	for i := 0; i < MaxBuffers; i++ {
+	// up-front reserve memory for the essential buffers (1 input, 1 output + the reader buffer)
+	if err := mp.memoryManager.ReserveMemory(3*BufferSize, 255); err != nil {
+		return nil, err
+	}
+
+	mp.reservedMemory += 3 * BufferSize
+	bufs := 1
+
+	// reserve some more memory for buffers if possible
+	for i := 1; i < MaxBuffers; i++ {
 		var prio uint8
-		switch bufs {
-		case 0:
-			prio = 255
-		case 1:
+		if bufs < 2 {
 			prio = 192
-		default:
+		} else {
 			prio = 128
 		}
-		if err = mp.memoryManager.ReserveMemory(2*MaxMessageSize, prio); err != nil {
+
+		if err := mp.memoryManager.ReserveMemory(2*BufferSize, prio); err != nil {
 			break
 		}
-		mp.reservedMemory += 2 * MaxMessageSize
+		mp.reservedMemory += 2 * BufferSize
```

Comment:
Where does the 2 come from? I'm pretty sure we're only using 1x. We shouldn't be accounting for the buffer passed by the _user_.

Reply:
one for input and one for output, we reserve 4+4 buffers.

Reply:
added a comment about this.